### DISCUSSION TITLE
TensorFlow with Google Coral.ai

### DISCUSSION THREAD

**Discussion by grasponcrypto** [immich-app/immich#73] [General]
Wondering if there is a way to utilize the coral.ai Edge TPU for facial/object recognition.

**Comment by alextran1502** [comment]
I am working on both features for local hosted services. Once that done, I will be working on an option for external services to be used for those machine learning services since a lot of people has their own model trained already 😃 I will keep your suggestion in mind

**Reply by Schluggi** [reply]
Any updates? :)

**Reply by mertalev** [reply]
The model engine we use, ONNX Runtime, doesn't currently support Coral ([relevant issue](https://github.com/microsoft/onnxruntime/issues/10248)). I'm happy to add it if/when this becomes an option.

**Comment by afrojewelz** [comment]
looking forward to see that outcome

**Comment by ptichalouf** [comment]
I'm here if U need some tester :)

**Comment by KillahB33** [comment]
Excited to see the transition to python for the ai piece which hopefully means that this will be coming soon enough!

**Comment by pdarcos** [comment]
I guess there is still no news on this from upstream?

**Reply by bo0tzz** [reply]
You would need to check the linked upstream issue for that.

**Comment by nikefate** [comment]
Feature Request for ONNX Runtime is closed.  Is there any possibility of obtaining support for Coral in another way?

**Reply by bo0tzz** [reply]
I think you're looking at the wrong upstream issue, it's still open.

**Comment by bo0tzz** [comment]
We investigated this again recently and the Coral is just not a capable enough device to run the models we use for Immich, so I'm closing this feature request.

**Reply by ncridlig** [reply]
Just followed this thread after acquiring a Coral from a buddy. Thank you for investigating the possibility!

**Reply by sa7mon** [reply]
> We investigated this again recently and the Coral is just not a capable enough device to run the models we use for Immich, so I'm closing this feature request.

Would you mind providing more details about why it's not capable enough? Has anything changed in 2 years?

**Reply by mertalev** [reply]
It has too little onboard memory to perform well for the model size. Even if it did (it doesn't), 4 TOPS for this day and age is very slow. It's simply dated hardware.

**Comment by Timi7007** [comment]
ONNX seems to support Hailo modules, like the new Raspberry AI module, though. Would this be an option?

**Reply by bo0tzz** [reply]
The main problem with these small TPUs is that they just don't have anywhere near enough RAM for the models that Immich uses. I can't find any actual numbers for Hailo, but it seems like it's just as far off as the Coral.

**Reply by mertalev** [reply]
I think NPUs just use normal RAM like iGPUs, so it may be different here.

**Reply by bo0tzz** [reply]
Hard numbers don't seem to exist, but [a comment on hackernews](https://news.ycombinator.com/item?id=40533687) guesses based on the models listed that it's probably some tens of MBs.

**Reply by chelming** [reply]
Halio-8L can run TensorFlow, TensorFlow Lite, Keras, PyTorch & ONNX models, where the Coral only does TFlite (and might max out at 6mb model size?).

I don't see any reference to a maximum model size for the Halio-8L and whether it needs to load the whole model into its **_own_** memory. There's not a lot of documentation or spec information for them.
![image](https://github.com/immich-app/immich/assets/7746625/26ec2f79-7a15-411d-bda0-dc83208fbc22)

**Reply by mertalev** [reply]
They have a CLIP section in their [examples](https://github.com/hailo-ai/hailo-rpi5-examples) (coming soon ™).

**Reply by d9media** [reply]
> ONNX seems to support Hailo modules, like the new Raspberry AI module, though. Would this be an option?

Hey guys, I've got a RPI5 laying around (8gb) and looking into getting the AI Kit (Hailo-8 M.2 PCIe). What's the current state on this? Will I be able to utilize the combo to run the ml container? Exciting stuff. I follow Immich for some time now and progress has been really impressive.

**Reply by mertalev** [reply]
I recently got a Hailo module for testing, but haven't started implementing support for it yet. I'll have more time to work on this in a few weeks.

**Reply by d9media** [reply]
> eks.

Awesome! Is there anything in this regard you could use help with? I'm reasonable familiar with docker and microservices but not really an expert in ML. If there's anything i could contribute I'd be glad to. I feel like like this setup could have some potential.

**Reply by mertalev** [reply]
I think the work will mainly be in exporting the models to Hailo format and making the machine learning service able to use them. I've been meaning to make the model exporting process more automated, so that's something I could use help with. There are several scripts for this that I have to run in the right order, some for each model, and with the Hailo models in the loop it'll get more tedious.

A tricky thing with hardware acceleration is also that it sometimes works on one machine but not on others. After I get things working locally, it'd be great if I could get a sanity check that it actually works.

**Reply by d9media** [reply]
@mertalev Is this a good pick going forward you reckon? https://buyzero.de/products/raspberry-pi-ai-hat-13-26-tops?variant=45993772089611 Apparently it's their successor of the original AI Kit and now based on Hailo 8L, and the larger on based on Hailo 8.

**Reply by mertalev** [reply]
Oh nice, I didn't know they came out with new versions of the kit. The 26 TOPS version seems especially interesting.

**Comment by zxcOvosh** [comment]
Pls Tell me how things are going with hailo 8h/google coral in immich now

**Reply by ptkambo** [reply]
Yes please :)

**Reply by USAemulators** [reply]
yes i need coral on immich too. just like it helped cpu on qumagie

**Reply by mertalev** [reply]
We won't be supporting Coral devices because they're too weak for the models we use. Hailo chips are more interesting, but there's no active work to support them at the moment and I don't anticipate there to be in the near future unless someone would like to contribute it.
